moveit_core/robot_model/src/joint_model_group.cpp

```diff
@@ -97,188 +101,239 @@ bool jointPrecedes(const JointModel* a, const JointModel* b)
+                 start_index_view)
+        | views::transform(tuple_to_pair)
+    )
+    | ranges::to<std::map>();
+  active_joint_models_bounds_ = active_joint_model_vector_
+    | views::transform(get_variable_bounds_addr)
+    | ranges::to<std::vector>();
+  variable_index_list_ =
+    views::zip(not_fixed_view | views::transform(get_first_variable_index),
+               not_fixed_view | views::transform(get_variable_names_size))
+    | views::transform([](const auto& e) {
+     const auto& [first_index, names_size] = e;
+     return views::iota(first_index, first_index + names_size);
+    })
+    | views::join
+    | ranges::to<std::vector>();
+  active_joint_model_start_index_ =
+    views::zip(start_index_view, not_fixed_view)
+    | views::remove_if([](const auto& e) { return is_mimic_joint(std::get<1>(e)); })
+    | views::transform([](const auto& e) { return std::get<0>(e); })
+    | ranges::to<std::vector>();
+  variable_count_ = ranges::accumulate(not_fixed_view | views::transform(get_variable_count), 0U);
+  active_variable_count_ = ranges::accumulate(not_fixed_view | views::transform(get_variable_count), 0U);
+  is_single_dof_ = ranges::none_of(not_fixed_view | views::transform(get_variable_count),
+    [](const auto& e) { return e > 1; });
 
   // now we need to find all the set of joints within this group
   // that root distinct subtrees
-  for (const JointModel* active_joint_model : active_joint_model_vector_)
-  {
-    // if we find that an ancestor is also in the group, then the joint is not a root
-    if (!includesParent(active_joint_model, this))
-      joint_roots_.push_back(active_joint_model);
-  }
+  joint_roots_ = active_joint_model_vector_
+    | views::filter([this](const auto& jm) {
+      return !includesParent(jm, this);
+    })
+    | ranges::to<std::vector>();
 
   // when updating this group within a state, it is useful to know
   // if the full state of a group is contiguous within the full state of the robot
-  if (variable_index_list_.empty())
-    is_contiguous_index_list_ = false;
-  else
-    for (std::size_t i = 1; i < variable_index_list_.size(); ++i)
-      if (variable_index_list_[i] != variable_index_list_[i - 1] + 1)
-      {
-        is_contiguous_index_list_ = false;
-        break;
-      }
-
-  // when updating/sampling a group state only, only mimic joints that have their parent within the group get updated.
-  for (const JointModel* mimic_joint : mimic_joints_)
-    // if the joint we mimic is also in this group, we will need to do updates when sampling
-    if (hasJointModel(mimic_joint->getMimic()->getName()))
-    {
-      int src = joint_variables_index_map_[mimic_joint->getMimic()->getName()];
-      int dest = joint_variables_index_map_[mimic_joint->getName()];
-      GroupMimicUpdate mu(src, dest, mimic_joint->getMimicFactor(), mimic_joint->getMimicOffset());
-      group_mimic_update_.push_back(mu);
-    }
+  const auto is_ascending = [](const auto& sub_view) {
```

Comment:
Should this be renamed? This function implies that any value greater would qualify, but this check to see if it is *exactly* the next one

Reply:
how about `is_contiguous` ?  I renamed it to that.